### PR TITLE
Improvements to the help system

### PR DESCRIPTION
**Pull request by jlstevens**
This PR implements some improvements to the help system, suggested in #331.

Supports filtering of parameters using regular expressions as well as highlight. For instance:

```python
hv.help(hv.Image, pattern='bounds')
```

image
 
```python
hv.help(hv.Image, visualization=False, pattern='bounds')
```
image

It seems to work well even though I needed to employ some clever tricks to filter parameters. Now I'll look into addressing issue #330 in this PR as well..

### CODE REVIEW

**Comment by jlstevens** [comment]
I've now added a recursive option to address #330. This option will only work when given a nested object and will not do anything is a class is specified instead.

Once the tests pass, I think this is ready to merge.

**Comment by philippjfr** [comment]
Looks good. Great to have a better help system. Merging.

**Comment by maxalbert** [comment]
Great, this seems to work well. Many thanks!

I wonder whether it would be worth setting `recursive=True` by default to increase discoverability of options for composite objects (but I don't have a strong opinion on it).

It would also be good to document this new functionality in one of the tutorials (maybe the "Options" tutorial, since it already illustrates the `help` command?).

**Comment by jlstevens** [comment]
> I wonder whether it would be worth setting recursive=True by default to increase discoverability of options for composite objects

Not a bad suggestion! I might do that after the website has been updated as changing it immediately will probably break notebook tests. 

> It would also be good to document this new functionality in one of the tutorials (maybe the "Options" tutorial, since it already illustrates the help command?).

Yes, that is definitely something I will do now. I can change Markdown text much more easily than output generated by code. Thank you for reminding me to do this!

**Comment by jlstevens** [comment]
I've added a brief description of these parameters in commit 04f77e4 to the Options tutorial.

@philippjfr Maybe we should use the pattern argument in the actual example? This would mean updating the test data but I see three advantages:

1. Even ``hv.help(image)`` is generating too much output in the tutorial, making it hard to follow.
2. It would mean we are less likely to have to update the test data when parameter documentation is updated.
3. It would give a concrete example of ``pattern`` in use and why it is helpful.

What do you think? To be concrete, I'm considering showing the output of ``hv.help(image, pattern='bounds')``...

**Comment by jbednar** [comment]
Sounds good to me.

**Comment by maxalbert** [comment]
+1

**Comment by jlstevens** [comment]
I've gone and implemented this suggestion in 5ed9eda.
